// node-red-contrib-re-postgres/lib/pg.js
var pg = require('pg');
var named = require('node-postgres-named');

module.exports=function(RED){

	function PostgresDatabaseNode(n) {
		RED.nodes.createNode(this, n);
		this.hostname = n.hostname;
		this.port = n.port;
		this.db = n.db;
		this.ssl = n.ssl;

		var credentials = this.credentials;
		if (credentials) {
			this.user = credentials.user;
			this.password = credentials.password;
		}
	}
	
	RED.nodes.registerType("postgresdb", PostgresDatabaseNode, {
		credentials: {
			user: {type: "text"},
			password: {type: "password"}
		}
	});


	function PostgresNode(n) {
		RED.nodes.createNode(this, n);
	
		var node = this;
	
		node.topic = n.topic;
		node.postgresdb = n.postgresdb;
		node.postgresConfig = RED.nodes.getNode(this.postgresdb);
		node.sqlquery = n.sqlquery;
		node.output = n.output;
		node.perrow = n.perrow;
	
		if (node.postgresConfig) {
	
			var connectionConfig = {
				user: node.postgresConfig.user,
				password: node.postgresConfig.password,
				host: node.postgresConfig.hostname,
				port: node.postgresConfig.port,
				database: node.postgresConfig.db,
				ssl: node.postgresConfig.ssl
			};
	
			var handleError = function(err, msg) {
				msg.error=err;
				node.error(err,msg);
				console.log(err);
				console.log(msg.payload);
				console.log(msg.queryParameters);
			};
				
			node.on('input', function(msg) {
				pg.connect(connectionConfig, function(err, client, done) {
					if (err) {
						handleError(err, msg);
					} else {
						named.patch(client);
				
						if (!!!msg.queryParameters)
							msg.queryParameters = {};
						
						var q=client.query(
							msg.payload,
							msg.queryParameters,
							function(err, results) {
								done();
								if (err) {
									handleError(err, msg);
								} else {
									if (node.output && !node.perrow) {
										msg.payload = results.rows;
										node.send(msg);
									}
								}
							}
						);
						if (!!node.perrow)
							q.on("row",function(row,res){
								node.send(Object.assign(Object.assign({},msg),{payload:row}));
							})
					}
				});
			});
		} else {
			this.error("missing postgres configuration");
		}

		this.on("close", function() {
			if (node.clientdb) node.clientdb.end();
		});
	}

	RED.nodes.registerType("postgres", PostgresNode);
}
